### PR TITLE
feat(provider/gae): Specify artifact in GAE deploy from GCS.

### PR DESCRIPTION
**Pull request by jtk54**
Also select between git and GCS during a GAE deploy, and select a storage account accordingly.

### CODE REVIEW

**Comment by jtk54** [comment]
Don't merge yet -- there's an issue with editing deploy config right now.

**Comment by skim1420** [comment]
Please add screenshots for UI work.

**Comment by jtk54** [comment]
image

**Comment by jtk54** [comment]
Fixed up editing the deploy config, should be functional now.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts [hunk 1]
```diff
@@ -30,8 +37,47 @@ class AppengineServerGroupBasicSettingsCtrl implements IController {
...
+  public isGcsSourceRepo(): boolean {
+    return this.$scope.command.sourceRepositoryType === AppengineSourceRepositoryType.GCS;
+  }
+
+  public isArtifactSource(): boolean {
+    return this.$scope.command.sourceRepositoryType === AppengineSourceRepositoryType.ARTIFACT;
+  }
+
+  public setFromArtifact(): void {
+    this.$scope.command.fromArtifact = this.isArtifactSource()
```
Comment (danielpeach) [inline app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: Missing a semicolon.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html [hunk 2]
```diff
@@ -59,63 +59,128 @@
       </div>
     </div>
 
-    <div class="form-group">
-      <div class="col-md-3 sm-label-right">
-        Git Repository URL <help-field class="help-field-absolute" key="appengine.serverGroup.repositoryUrl"></help-field>
+    <div class="form-group row">
+      <label class="col-md-3 sm-label-right">Source Repository Type</label>
```
Comment (danielpeach) [inline app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: s/Source Repository Type/Source Type

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html [hunk 3]
```diff
@@ -59,63 +59,128 @@
...
+      <div class="col-md-9">
+        <select class="input-sm"
```
Comment (danielpeach) [inline app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: For a short fixed list of options, it's better to use radio buttons instead of a select. It's one click instead of two to select an option, and you can see all the options immediately.
See https://github.com/spinnaker/deck/blob/master/app/scripts/modules/appengine/loadBalancer/configure/wizard/stageAllocationConfigurationRow.component.html#L35-L50

File: settings.js [hunk 4]
```diff
@@ -140,12 +139,11 @@ window.spinnakerSettings = {
     netflixMode: netflixMode,
     chaosMonkey: chaosEnabled,
     // whether stages affecting infrastructure (like "Create Load Balancer") should be enabled or not
-    infrastructureStages: infrastructureEnabled,
+    infrastructureStages: process.env.INFRA_STAGES === 'enabled',
```
Comment (danielpeach) [inline settings.js]: This feels like an unrelated change...

File: app/scripts/modules/core/src/storage/storageAccount.service.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import {module} from 'angular';
```
Comment (danielpeach) [inline app/scripts/modules/core/src/storage/storageAccount.service.ts]: Space around the imports here and below.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html [hunk 6]
```diff
@@ -59,63 +59,128 @@
...
       </div>
     </div>
 
-    <div class="form-group">
-      <div class="col-md-3 sm-label-right">
-        Branch <help-field key="appengine.serverGroup.branch"></help-field>
+    <div ng-if="basicSettingsCtrl.isGcsSourceRepo()">
+      <div class="form-group">
+        <div class="col-md-3 sm-label-right">
+          GCS URL <help-field class="help-field-absolute" key="appengine.serverGroup.repositoryUrl"></help-field>
```
Comment (danielpeach) [inline app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: The help text here (from `appengine.serverGroup.repositoryUrl`) only makes sense for git repos - I would make a separate help field.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html (hunk 3 above)
Comment (danielpeach) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: Also, you should be careful about remains on the command when you change `sourceType`.

For example, if I am using a `git` type and have a `sourceRepository` of `https://myGitRepo.git`, that value is going to show up inside the `GCS URL` input field if I change `sourceType` to `gcs`. The easiest thing is to just clear `sourceRepository` when you change `sourceType`.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts [hunk 7]
```diff
@@ -1,22 +1,29 @@
...
+import { AppengineSourceRepositoryType, IAppengineServerGroupCommand } from '../serverGroupCommandBuilder.service';
 
 interface IAppengineBasicSettingsScope extends IScope {
   command: IAppengineServerGroupCommand;
 }
 
 class AppengineServerGroupBasicSettingsCtrl implements IController {
+
+  public sourceRepositoryTypes: string[] = [];
+  public appEngineStorageAccounts: string[];
```
[the remark this answers is not in the document]
Comment (danielpeach) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: They're not scoped to App Engine at all, are they? I would just say `storageAccounts`.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html [hunk 8]
```diff
@@ -59,63 +59,128 @@
...
     </div>
 
-    <div class="form-group">
-      <div class="col-md-3 sm-label-right">
-        Git Credential Type <help-field class="help-field-absolute" key="appengine.serverGroup.gitCredentialType"></help-field>
+    <div ng-if="basicSettingsCtrl.isArtifactSource()">
+      <div class="form-group">
+        <label class=" col-md-3 sm-label-right">Expected Artifact</label>
+        <div class="col-md-7">
+          <ui-select ng-if="command.backingData.expectedArtifacts.length"
```
Comment (danielpeach) [inline app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: Having an empty dropdown here is fine - or you could have a message that says something like `You have not configured any expected artifacts` - but blank space is confusing.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html [hunk 9]
```diff
@@ -59,63 +59,128 @@
...
+            </ui-select-choices>
+          </ui-select>
+        </div>
+
+        <div class="col-md-7 col-md-offset-3" ng-if="command.viewState.mode === 'createPipeline' || command.viewState.mode === 'editPipeline'">
         <span class="pull-right small" ng-if="!command.fromTrigger">
           <a href ng-click="basicSettingsCtrl.toggleResolveViaTrigger()">Resolve via trigger</a>
         </span>
-        <span class="pull-right small" ng-if="command.fromTrigger">
+          <span class="pull-right small" ng-if="command.fromTrigger">
```
Comment (danielpeach) [inline app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: Extra indent

**Comment by danielpeach** [comment]
UX-wise -

I would change the `Source Types` to radio buttons labeled `git` and `gcs`.

Under `gcs`, 
- I would put `Storage Account` first, since that is going to be configured regardless. Would also make it span the same length as the other fields (e.g., `Account`).
- Rather than `Deploy From Expected Artifact` (which you have in your screenshots, but I think is gone now), I would have 

```
Resolve URL     [radio] via text input [radio] via pipeline artifact
```

and have an unlabeled input (for the url) or an unlabeled select dropdown (for the artifacts).

File: app/scripts/modules/appengine/serverGroup/configure/serverGroupCommandBuilder.service.ts [hunk 10]
```diff
@@ -44,7 +45,17 @@ export interface IAppengineServerGroupCommand {
   fromTrigger?: boolean;
   trigger?: IAppengineGitTrigger | IAppengineJenkinsTrigger;
   gitCredentialType?: GitCredentialType;
+  storageAccountName?: string; // GCS only
   interestingHealthProviderNames: string[];
+  fromArtifact: boolean;
+  expectedArtifact: IExpectedArtifact;
```
[the remark this answers is not in the document]
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/serverGroupCommandBuilder.service.ts]: Yes, we have a bug open for this refactor in Deck, Echo, and Orca.

File: app/scripts/modules/appengine/serverGroup/configure/serverGroupCommandBuilder.service.ts [hunk 11]
```diff
@@ -44,7 +45,17 @@ export interface IAppengineServerGroupCommand {
...
+  sourceRepositoryType: AppengineSourceRepositoryType;
+}
+
+export enum AppengineSourceRepositoryType {
```
[the remark this answers is not in the document]
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/serverGroupCommandBuilder.service.ts]: Done.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts [hunk 12]
```diff
@@ -1,22 +1,29 @@
 import { extend, IController, IControllerService, IScope, module } from 'angular';
 import { StateService } from '@uirouter/angularjs';
+import { STORAGE_ACCOUNT_SERVICE, StorageAccountService } from 'core/storage/storageAccount.service';
 import { set } from 'lodash';
+import { copy } from 'angular';
```
[the remark this answers is not in the document]
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: Done.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: At this point they are scoped only to appengine, but that is changing in the future. Renamed.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts [hunk 13]
```diff
@@ -30,8 +37,47 @@ class AppengineServerGroupBasicSettingsCtrl implements IController {
     if (!this.$scope.command.gitCredentialType) {
       this.onAccountChange();
     }
+
+    for (let type in AppengineSourceRepositoryType) {
```
[the remark this answers is not in the document]
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: Done.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts [hunk 14]
```diff
@@ -30,8 +37,47 @@ class AppengineServerGroupBasicSettingsCtrl implements IController {
...
+      // Linter complains if the for is not filtered with an if.
+      if (type) {
+        this.sourceRepositoryTypes.push(AppengineSourceRepositoryType[type]);
+      }
+    }
+
+    storageAccountService.getStorageAccounts().then(storageAccounts => {
```
[the remark this answers is not in the document]
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: Done.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts (hunk 1 above)
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: Whoops.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html (hunk 2 above)
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: Done.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html (hunk 9 above)
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: Intellij black magic.

File: app/scripts/modules/core/src/storage/storageAccount.service.ts (hunk 5 above)
Comment (jtk54) [reply app/scripts/modules/core/src/storage/storageAccount.service.ts]: Done.

File: settings.js (hunk 4 above)
Comment (jtk54) [reply settings.js]: settings.js isn't included in this change, I'll revert that file.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts [hunk 15]
```diff
@@ -1,17 +1,20 @@
...
-import { IAppengineServerGroupCommand } from '../serverGroupCommandBuilder.service';
+import { AppengineSourceType, IAppengineServerGroupCommand } from '../serverGroupCommandBuilder.service';
 
 interface IAppengineBasicSettingsScope extends IScope {
   command: IAppengineServerGroupCommand;
 }
 
 class AppengineServerGroupBasicSettingsCtrl implements IController {
+
+  public readonly sourceTypes: string[] = Object.keys(AppengineSourceType).map(k => AppengineSourceType[k as any]);
```
Comment (danielpeach) [inline app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: Do you need this any more? I'm not seeing it anywhere else.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html [hunk 16]
```diff
@@ -59,63 +59,144 @@
...
+            <ui-select-match>{{ basicSettingsCtrl.summarizeExpectedArtifact(command.expectedArtifact) }}
+            </ui-select-match>
+            <ui-select-choices repeat="expected in command.backingData.expectedArtifacts">
+              <span>{{ basicSettingsCtrl.summarizeExpectedArtifact(expected) }}</span>
+            </ui-select-choices>
+          </ui-select>
+        </div>
+      </div>
+
+      <div class="form-group" ng-if="!!!command.fromArtifact">
```
Comment (danielpeach) [inline app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: Not not not?

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts [hunk 17]
```diff
@@ -32,6 +35,26 @@ class AppengineServerGroupBasicSettingsCtrl implements IController {
     }
   }
 
+  public isGitSource(): boolean {
+    return this.$scope.command.sourceType === AppengineSourceType.GIT as string;
```
Comment (danielpeach) [inline app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: I am very surprised you have to cast this but I believe you.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts (hunk 15 above)
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: Removed.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts (hunk 17 above)
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.controller.ts]: Removed.

File: app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html (hunk 16 above)
Comment (jtk54) [reply app/scripts/modules/appengine/serverGroup/configure/wizard/basicSettings.html]: I think I tried to use `!!` to cast this at some point and forgot to change it back.
